ghproxy/ghcache/coalesce.go

```diff
@@ -0,0 +1,108 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ghcache
+
+import (
+	"bufio"
+	"bytes"
+	"net/http"
+	"net/http/httputil"
+	"sync"
+)
+
+// requestCoalescer allows concurrent requests for the same URI to share a
+// single upstream request and response.
+type requestCoalescer struct {
+	sync.Mutex
+	keys map[string]*responseWaiter
+
+	delegate http.RoundTripper
+}
+
+type responseWaiter struct {
+	*sync.Cond
+
+	waiting bool
+	resp    []byte
+	err     error
+}
+
+// RoundTrip coalesces concurrent GET requests for the same URI by blocking
+// the later requests until the first request returns and then sharing the
+// response between all requests.
+//
+// Notes: Deadlock shouldn't be possible because the map lock is always
+// acquired before responseWaiter lock if both locks are to be held and we
+// never hold multiple responseWaiter locks.
+func (r *requestCoalescer) RoundTrip(req *http.Request) (*http.Response, error) {
+	// Only coalesce GET requests
+	if req.Method != http.MethodGet {
+		return r.delegate.RoundTrip(req)
+	}
+
+	key := req.URL.String()
+	r.Lock()
+	waiter, ok := r.keys[key]
+	if ok {
+		// Earlier request in flight. Wait for it's response.
+		defer req.Body.Close() // Since we won't pass the request we must close it.
+		waiter.L.Lock()
+		r.Unlock()
+		waiter.waiting = true
+		waiter.Wait()
```

Comment:
Do we need to worry about this from GoDoc?

> Because c.L is not locked when Wait first resumes, the caller typically cannot assume that the condition is true when Wait returns. Instead, the caller should Wait in a loop.

Reply:
No it doesn't apply to this use case because the condition we are waiting for remains true once it becomes true. 
I added a comment.